RPC: creates possibility to preserve labels on importprivkey

marcoagner opened the pull request:
Closes #13087.

As discussed in the issue, this is a feature request instead of a bug report since the behaviour was as intended (i.e. label with default: `''`). With this, the old behaviour is kept while the possibility to achieve the preservation of labels, as expected in the open issue, is added.

promag commented:
There is no need for a new parameter, check if label is null instead.

promag commented:
Concept ACK, a test would be nice.

marcoagner commented:
Thank you, @promag.
I added some functional tests for this specific behavior between importaddress and importprivkey.
If everything is alright, I'll squash and rebase as necessary.

promag commented on src/wallet/rpcdump.cpp:
```diff
@@ -151,9 +151,12 @@ UniValue importprivkey(const JSONRPCRequest& request)
+            // We don't know which corresponding address will be used; label all new addresses and preserve the rest
             for (const auto& dest : GetAllDestinationsForKey(pubkey)) {
-                pwallet->SetAddressBook(dest, strLabel, "receive");
+                bool has_label = pwallet->mapAddressBook.find(dest) != pwallet->mapAddressBook.end() && !pwallet->mapAddressBook[dest].name.empty();
```
I was thinking in something like
```cpp
if (request.params[1].isNull()) {
    const auto& it = pwallet->mapAddressBook.find(dest);
    if (it != pwallet->mapAddressBook.end()) strLabel = it->second;
}
pwallet->SetAddressBook(dest, strLabel, "receive");
```

marcoagner replied:
Hm, I see. Yeah.. this behavior seems better. Thanks!
Updating this and the test.

marcoagner commented:
Updated with the right behavior.
I will squash and rebase it as necessary in the morning, thanks.

promag reviewed (commented):
Update PR title.

harding commented:
Tested ACK 74877f3b10072e1975c931bb296212e008cd8c58 .  This is exactly the behavior I desired, thanks @marcoagner!

promag commented on src/wallet/rpcdump.cpp:
```diff
@@ -151,8 +151,15 @@ UniValue importprivkey(const JSONRPCRequest& request)
             for (const auto& dest : GetAllDestinationsForKey(pubkey)) {
+                if (request.params[1].isNull()) {
+                    const auto& it = pwallet->mapAddressBook.find(dest);
+                    if (it != pwallet->mapAddressBook.end()) strLabel = it->second.name;
```
Good catch! Suggestion:
```cpp
if (it == pwallet->mapAddressBook.end()) {
    pwallet->SetAddressBook(dest, strLabel, "receive")
} else {
    pwallet->SetAddressBook(dest, it->second.name, "receive")
}
```

promag replied:
There should be a test ensuring the correct behaviour.

marcoagner replied:
Yes, makes sense. I will update it and add a test for the correct behaviour as soon as I can.

marcoagner commented:
I'm back to this. The new commit keeps the the already tested behaviour (passing tests added on this PR et al) and passes new test for the multiple destinations bug spotted by @Empact.

luke-jr commented:
This still changes behaviour: the default is now `null` (ie, current label) instead of `""`.

I don't particularly care what the default is (`null` makes more sense), but the change should be documented and noted in release notes.

promag commented:
Agree with @luke-jr regarding the behaviour change.

marcoagner commented:
I added a `release-notes-pr13381.md` file documenting the change.
Let me know if I missed something or should document this somewhere else.
Thanks!

harding commented:
@marcoagner Thanks for adding the release notes!  I thought maybe the behavior change could be described more simply with the addition of a few examples; see below from some proposed text.  I also changed the heading so that it's clear that only a small part of the importprivkey API is being changed (I also removed the `code style` from the heading as those don't display well in headings on parts of BitcoinCore.org, but that's a local problem so feel free to add back the backtics if you like them).

```markdown
RPC importprivkey: new label behavior
-------------------------------------

Previously, `importprivkey`automatically added the default empty label
("") to all addresses associated with imported private keys.  Now it
defaults to using any existing label for those addresses.  For example:

- Old behavior: you import a watch-only address with the label "cold
  wallet".  Later, you import the corresponding private key using the
  default settings.  The address's label is changed from "cold wallet"
  to "".

- New behavior: you import a watch-only address with the label "cold
  wallet".  Later, you import the corresponding private key using the
  default settings.  The address's label remains "cold wallet".

In both the previous and current case, if you directly specify a label
during the import, that label will override whatever previous label the
addresses may have had.  Also in both cases, if none of the addresses
previously had a label, they will still receive the default empty label
("").  Examples:

- You import a watch-only address with the label "temporary".  Later you
  import the corresponding private key with the label "final".  The
  address's label will be changed to "final".

- You import a private key for an address that was not previously in the
  wallet.  It's addresses will receive the default empty label ("").
```

marcoagner commented:
@harding That's much better, thanks.
I'm changing to that on the next commit with some small differences.
Here's the diff so it's easier to spot my edits:
```diff
4,5c4,5
< Previously, `importprivkey`automatically added the default empty label
< ("") to all addresses associated with imported private keys.  Now it
---
> Previously, `importprivkey` automatically added the default empty label
> ("") to all addresses associated with the imported private keys.  Now it
21c21
< ("").  Examples:
---
> ("") as default.  Examples:
28c28,29
<   wallet.  It's addresses will receive the default empty label ("").
---
>   wallet using the default settings.  Its addresses will receive the
>   default empty label ("").
```
Seems good?

harding commented:
@marcoagner LGTM, although the two uses of 'default' in this clause is a tiny bit awkward IMO (but, it's clear, which is all that really matters):

> they will still receive the default empty label ("") as default.

Maybe that second-to-last sentence could have its adverbial phrase "using the default settings" moved next to the verb it modifies, "import", e.g.:

> You use the default settings to import a private key for an address that was not previously in the wallet.

Thanks for updating, and thanks especially for catching my misuse of the apostrophe in the possessive its.  :man_facepalming:

marcoagner commented:
@harding I changed both parts pointed in your last comment. Thanks for the attention and help with this; I think we've now nailed this part. :)

marcoagner commented:
Could anybody review 6115b5231954a38c400fa4808627dbb8e239ea68, please? One month in the freezer and is (hopefully) simple. :)

sipa commented on src/wallet/rpcdump.cpp:
```diff
@@ -151,9 +151,17 @@ UniValue importprivkey(const JSONRPCRequest& request)
+            // label was passed.
             for (const auto& dest : GetAllDestinationsForKey(pubkey)) {
-                pwallet->SetAddressBook(dest, strLabel, "receive");
+                const auto& it = pwallet->mapAddressBook.find(dest);
```
Nit: It's slightly confusing to use a reference type here; `find` returns an iterator by value.

sipa reviewed (commented):
utACK 6115b5231954a38c400fa4808627dbb8e239ea68. Didn't review the tests.

Non-blocking nit only.

promag commented on test/functional/wallet_import_with_label.py:
```diff
@@ -0,0 +1,126 @@
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Test the behavior of RPC importprivkey on set and unset labels of
+watch-only addresses.
```
Drop `of watch-only`?

promag commented on test/functional/wallet_import_with_label.py:
```diff
@@ -0,0 +1,126 @@
+        self.num_nodes = 2
+        self.setup_clean_chain = True
+
+    def setup_network(self):
```
Can be removed.

sipa commented on src/wallet/rpcdump.cpp:
```diff
@@ -151,9 +151,16 @@ UniValue importprivkey(const JSONRPCRequest& request)
-                pwallet->SetAddressBook(dest, strLabel, "receive");
+                if (!request.params[1].isNull()) {
+                    pwallet->SetAddressBook(dest, strLabel, "receive");
+                } else if (pwallet->mapAddressBook.count(dest) == 0) {
```
Given that the branches for the two `if` statements are identical, why write it as a single `if (!request.params[1].isNull() || pwallet->mapAddressBook.count(dest)) {` ?

marcoagner replied:
Just to try to make it clear in the code that the `request.params[1]` has precedence. Bad idea?

marcoagner replied:
No reason besides it since AFAIK putting the look up as second condition in the `||` statement will not call it if there is a `params[1]` in the first condition.

sipa reviewed (commented):
utACK 1d4fdb33349a60cac1926f83586826ba62b1a451

MarcoFalke commented on test/functional/wallet_import_with_label.py:
```diff
@@ -0,0 +1,122 @@
+        self.num_nodes = 2
+        self.setup_clean_chain = True
+
+    def run_test(self):
```
The skip-if-no-wallet might have to be added to this test?

marcoagner replied:
You're right. I was not aware, thanks.

MarcoFalke commented:
Lets see what travis says

marcohextor committed:
rpc: creates possibility to preserve labels on importprivkey

- changes importprivkey behavior to overwrite existent label if one
is passed and keep existing ones if no label is passed

- tests behavior of importprivkey on existing address labels and
different same key destination

marcoagner commented on test/functional/wallet_import_with_label.py:
```diff
@@ -0,0 +1,126 @@
+
+        bcaddress_assert = self.nodes[1].getaddressinfo(
+                address_assert4['embedded']['address']
+                )
```
Hm, nice. Didn't know this `black`, thanks. 
I was using Flake8. Besides the double-quotes for strings, it seems that `black` has a longer default line length, but I'm okay with (actually prefer) that. :)
I will just rebase, re-run tests, and provide a new commit.

marcoagner commented:
`wallet_basic.py` passed locally but seems to have failed on appveyor :thinking:

MeshCollider commented:
utACK https://github.com/bitcoin/bitcoin/pull/13381/commits/a6b5ec18ff39ef3ccd19ec0e6db9ae00602d8938

sipa commented:
utACK a6b5ec18ff39ef3ccd19ec0e6db9ae00602d8938

jonasschnelli commented:
utACK a6b5ec18ff39ef3ccd19ec0e6db9ae00602d8938
